src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+            get => GetValue(key);
+            set => SetValue(key, value);
+        }
+
+        public TValue this[int index]
+        {
+            get => GetAt(index);
+            set => SetAt(index, value);
+        }
+
+        public OrderedDictionary()
+            : this(0, null)
+        {
+        }
+
+        public OrderedDictionary(int capacity)
+            : this(capacity, null)
+        {
+        }
+
+        public OrderedDictionary(IEqualityComparer<TKey> comparer)
+            : this(0, comparer)
+        {
+        }
+
+        public OrderedDictionary(int capacity, IEqualityComparer<TKey> comparer)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+            if (capacity > 0)
+            {
+                Resize(HashHelpers.GetPrime(capacity));
+            }
+
+            if (comparer != EqualityComparer<TKey>.Default)
+            {
+                _comparer = comparer;
+            }
+        }
+
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection)
+            : this(collection, null)
+        {
+        }
+
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey> comparer)
+            : this((collection as ICollection<KeyValuePair<TKey, TValue>>)?.Count ?? 0, comparer)
+        {
+            if (collection == null)
+            {
+                throw new ArgumentNullException(nameof(collection));
+            }
+
+            foreach (KeyValuePair<TKey, TValue> pair in collection)
+            {
+                Add(pair.Key, pair.Value);
+            }
+        }
+
+        public void Add(TKey key, TValue value) => TryInsert(null, key, value, InsertionBehavior.ThrowOnExisting);
+
+        public void Clear()
+        {
+            if (Count > 0)
+            {
+                Array.Clear(_buckets, 0, _buckets.Length);
+                Array.Clear(_entries, 0, Count);
+                Count = 0;
+            }
+            ++_version;
+        }
+
+        public bool ContainsKey(TKey key) => IndexOf(key) >= 0;
+
+        public int EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            if (_entries.Length >= capacity)
+            {
+                return _entries.Length;
+            }
+            int newSize = HashHelpers.GetPrime(capacity);
+            Resize(newSize);
+            ++_version;
+            return newSize;
+        }
+
+        public TValue GetAt(int index) => GetAt(index, out _);
+
+        public TValue GetAt(int index, out TKey key)
+        {
+            if (index < 0 || index >= Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
```

Comment:
Please use only strings from resx (although we're not localizing, all the other code does)

eg `SR.ArgumentOutOfRange_Index`

Reply:
Same elsewhere

Reply:
Also should we use a ThrowHelper for collections in this experimental package as well? To minimize IL size?

https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/ThrowHelper.cs

Reply:
I don't think it's very important at this point.